proposals/required-members.md

```diff
@@ -263,42 +263,131 @@ init_argument_initializer
+    public new int Prop2 { get; }
+}
+```
+
+It is an error to hide a `required` member, as that member can no longer be set by a consumer.
+
+When overriding a `required` member, the `required` keyword must be included on the method signature. This is done so that if we ever want to allow
+unrequiring a property with an override in the future, we have design space to do so.
+
+Overrides are allowed to mark a member `required` where it was not `required` in the base type. A member so-marked is added to the required members
+list of derived type.
+
+### Metadata Representation
+
+The following 2 attributes are known to the C# compiler and required for this feature to function:
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct, AllowMultiple = false)]
+public sealed class RequiredMembersAttribute : Attribute
+{
+    public RequiredMembersAttribute(string[] members) => Members = members;
+    public string[] Members { get; }
+}
+
+[AttributeUsage(AttributeTargets.Constructor)]
+public sealed class NoRequiredMembersAttribute : Attribute
+{
+    public NoRequiredMembersAttribute() {}
+}
+```
+
+It is an error to manually apply `RequiredMembersAttribute` to a type.
+
+All members that are required by a type are listed in a `RequiredMembersAttribute` on the type. Members from any base type are not included, except
+when an override adds `required` to a member that was not `required` in the base type. As these members can only be fields or properties, duplicate
+names cannot exist.
+
+Any constructor in a type with `required` members that does not have `NoRequiredMembers` applied to it is marked with an `Obsolete` attribute with
+the string `"Types with required members are not supported in this version of your compiler"`, and the attribute is marked as an error, to prevent
+any older compilers from using these constructors. We don't use a `modreq` here because it is a goal to maintain binary compat: if the last `required`
+property was removed from a type, the compiler would no longer synthesize this `modreq`, which is a binary-breaking change and all consumers would
+need to be recompiled. A compiler that understands `required` members will ignore this obsolete attribute. Note that members can come from base types
+as well: even there are no new `required` members in the current type, if any base type has `required` members, this `Obsolete` attribute will be
+generated. If the constructor already has an `Obsolete` attribute, no additional attribute will be generated. If a by-ref-like type also has `required`
+members, the generated `Obsolete` attribute will reference the `required` error message, under the assumption that if a compiler is new enough to
+understand `required` properties, it will also understand by-ref-like-types.
+
+To build the full list of all `required` members for a given type T, including all base types, all the `RequiredMembersAttribute`s from T and its
```

Comment:
>T [](http://example.com/codeflow?start=66&length=1)

nit: may want to add back-ticks to the names in this sentence (`T`, `Tb`, `Ti`, `R1`, etc)
nit: "all" feels redundant in "the full list of all `required` members"

This is not a spec problem, but we'll want to test with invalid attributes on base type (`RequiredMembers("missing")` for instance). Will add a note to test plan.